File: baekjoon/버블_소트.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strings"
)

var count = 0
var nums = []int{}
var temp = []int{}

func mergeSort(left, mid, right int) {
	i := left
	j := mid + 1
	k := left

	for i <= mid && j <= right {
		if nums[i] <= nums[j] {
			temp[k] = nums[i]
			i++
		} else {
			temp[k] = nums[j]
			j++
			count += mid - i + 1
		}
		k++
	}

	for i <= mid {
		temp[k] = nums[i]
		i++
		k++
	}

	for j <= right {
		temp[k] = nums[j]
		j++
		k++
	}

	for i := left; i <= right; i++ {
		nums[i] = temp[i]
	}
}

func merge(left, right int) {
	if left < right {
		mid := (left + right) / 2
		merge(left, mid)
		merge(mid+1, right)
		mergeSort(left, mid, right)
	}
}

func main() {
	scanner := bufio.NewScanner(os.Stdin)

	var num int

	scanner.Scan()
	line := scanner.Text()
	fmt.Sscanf(line, "%d", &num)

	scanner.Scan()
	line = scanner.Text()
	splited := strings.Split(line, " ")
	for i := 0; i < len(splited); i++ {
		var n int
		fmt.Sscanf(splited[i], "%d", &n)
		nums = append(nums, n)
		temp = append(temp, 0)
	}

	merge(0, num-1)
	fmt.Println(count)
}

File: baekjoon/버블_소트.js
let [n, cases] = require("fs")
  .readFileSync("/dev/stdin")
  .toString()
  .trim()
  .split("\n");

n = Number(n);
cases = cases.split(" ").map(Number);

let count = 0;

const mergeSort = (left, mid, right) => {
  let i = left;
  let j = mid + 1;
  let k = left;

  while (i <= mid && j <= right) {
    if (cases[i] <= cases[j]) {
      temp[k++] = cases[i++];
    } else {
      temp[k++] = cases[j++];
      count += mid - i + 1;
    }
  }

  while (i <= mid) {
    temp[k++] = cases[i++];
  }

  while (j <= right) {
    temp[k++] = cases[j++];
  }

  for (let i = left; i <= right; i++) {
    cases[i] = temp[i];
  }
};

const merge = (left, right) => {
  if (left < right) {
    const mid = Math.floor((left + right) / 2);
    merge(left, mid);
    merge(mid + 1, right);
    mergeSort(left, mid, right);
  }
};

const temp = Array.from({ length: n }, () => 0);
merge(0, n - 1);
console.log(count);
